### DISCUSSION TITLE
portable far2m

### DISCUSSION THREAD

**Discussion by sorcodiv** [shmuz/far2m#22] [Ideas]
Здравствуйте!
Спасибо за [оперативный ответ](https://github.com/shmuz/far2l/issues/21) :)
Адаптировал [скрипт создания portable-сборки](https://github.com/unxed/far2l-deb/tree/master/portable) для far2l с lua.
[lxc-far2l-portable.zip](https://github.com/shmuz/far2l/files/10529302/lxc-far2l-portable.zip)
Основной файл для запуска сборки: lxc-far2l-portable.sh
Я проверял сборку в debian 11 в консоли (без установленного оконного менеджера).
Кратко скрипт делает следующее:
- устанавливает [lxc](https://ru.wikipedia.org/wiki/LXC);
- создает (и настраивает) новый контейнер с ubuntu 16.04;
- копирует в контейнер файлы luajit-2.0.patch, autonomizer.sh, make_far2l_portable_on_ubuntu_16_04.sh (в оригинальном скрипте все файлы скачиваются из репозитария автора, но я вносил в них изменения, необходимые для сборки);
luajit-2.0.patch добавляет возможность сборки с luajit-2.0. Если добавить изменения в основной код, то патч будет не нужен.
- внутри контейнера запускает make_far2l_portable_on_ubuntu_16_04.sh;
дальнейшие действия производятся внутри контейнера:
- устанавливаются нужные пакеты;
- скачивается актуальный код github.com/shmuz/far2l;
- применяется патч с версией luajit и патч "[«одиночное нажатие ESC» включены по умолчанию](https://github.com/unxed/far2l-deb/)";
- происходит сборка. Ограничение сохраняется: [отсутствует поддержка протокола SMB в NetRocks](https://github.com/unxed/far2l-deb/);
- после сборки копируются необходимые библиотеки, которые оригинальный скрипт не добавляет;
- запускается крипт, адаптирующий собранные бинарные файлы к "правильному" расположению нужных библиотек (в текущем каталоге), для этого используется утилита [patchelf](https://github.com/NixOS/patchelf);
- скачивается скрипт для [создания самораспаковывающегося архива](https://github.com/megastep/makeself), все нужные файлы для запуска far упаковывается в far2l_portable.run.

Запуск полученной сборки поверхностно проверил на дистрибутивах:
archlinux - в чистой консоли (без оконного менеджера) и в [putty4far2l](https://github.com/unxed/putty4far2l),
debian-11 - в чистой консоли (без оконного менеджера) и в [putty4far2l](https://github.com/unxed/putty4far2l),
manjaro-kde-live - в GUI,
ubuntu-22.04-desktop-live - в GUI.
Проверка совсем простая - запуск far и запуск в редакторе упрощенной версии скрипта [Duplicate Fighter](https://forum.farmanager.com/viewtopic.php?t=9900). Ошибок не обнаружил.

Автору оригинального скрипта создания portable-сборки я планирую чуть позже написать, возможно он захочет добавить себе поддержку far2l+lua.

**Comment by shmuz** [comment]
Спасибо, это очень интересно, буду постепенно разбираться.
Если будут ко мне вопросы - задавайте :)

**Comment by shmuz** [comment]
Есть вопросы:
1. Почему была выбрана именно Ubuntu 16 ?
2. Почему заложена сборка без WX, может быть можно добавить опцию комстроки, управляющую этим?
3. Если far2l будет работать как бы внутри Ubuntu 16, будет ли у него доступ ко всем папкам "родительской" ОС ?
4. В каком виде получается выходной "дистрибутив", это видимо один файл ?

Я многого ещё не знаю в мире Линукса, поэтому могу задавать и глупые вопросы :)

**Comment by sorcodiv** [comment]
Я не являюсь автором первоначального скрипта, брал его отсюда: https://github.com/unxed/far2l-deb/tree/master/portable
1. Предполагаю, что Ubuntu 16 выбрана для максимальной совместимости. Я для эксперимента попробовал собрать в Ubuntu 22, внутри контейнера far запускается, но за пределами контейнера получаю Segmentation fault. Я пока с причинами проблемы не разбирался.
2. Без WX - опять же предполагаю, что для совместимости. Опция "-DUSEWX=no" прописана в "make_far2l_portable_on_ubuntu_16_04.sh", я компилирую из консоли, поэтому у меня библиотека wxgtk не установлена.
3. Полученный файл far2l_portable.run запускается нативно, он не содержит в себе элементов виртуализации. Контейнер используется только для сборки, чтобы обеспечить одинаковость окружения независимо от установленной версии linux. После сборки итоговый файл копируется из контейнера на host-машину и используется без виртуализации.
Если говорить о доступе из host-системы (debian) к файловой системе внутри контейнера, то она находится по пути: /var/lib/lxc/far2l/rootfs/ (far2l - имя, заданное в скрипте lxc-far2l-portable.sh).
4. Выходной файл собирается с помощью скриптов https://github.com/megastep/makeself . Это по сути самораспаковывающийся tar.gz архив, который распаковывает себя во временный каталог и запускает ./far2l (в windows так умеет делать 7zip используя 7z.sfx). Файл far2l_portable.run можно распаковать плагином arclite (multiarc из far2l так не умеет). Фактически можно распаковать архив far2l_portable.run в любое место и запускать ./far2l - будет то же самое. Главное преимущество такой сборки - она запустится на "чистой" системе без необходимости ставить какие-либо библиотеки (например, lua), в том числе в live-окружении.

**Reply by shmuz** [reply]
Спасибо, со временем хотелось бы попробовать это всё на практике.
О WX я спросил потому что использую far2l исключительно с WX, т.к. в режиме TTY много клавиш и сочетаний недоступно, да и цвета по умолчанию ужасные. Насчёт цветов, это видимо исправимо, а насчёт клавиш - не уверен.

**Reply by Zeroes1** [reply]
О WX я спросил потому что использую far2l исключительно с WX, т.к. в режиме TTY много клавиш и сочетаний недоступно, 

А можно узнать какой перечень  хотя бы основной?

**Reply by shmuz** [reply]
Сейчас посмотрел - наткнулся только на `F10` неработающую. Все сочетания, естественно, не проверял.
Возможно, я писал про много недоступных клавиш до изменений, сделанных с тех пор.
А, вот ещё: `Ctrl/`

**Reply by Zeroes1** [reply]
> Сейчас посмотрел - наткнулся только на `F10` неработающую. Все сочетания, естественно, не проверял. Возможно, я писал про много недоступных клавиш до изменений, сделанных с тех пор. А, вот ещё: `Ctrl/`

хмм надо в GUI проверить по ssh по крайней мере проблем с этими нет.

**Reply by shmuz** [reply]
По SSH у меня F10 работает, Ctrl/ не работает, Ctrl. не работает (невозможно записать клавиатурный макрос).
Но может это потому, что у меня оригинальный (не патченый) PuTTY.

**Reply by Zeroes1** [reply]
> По SSH у меня F10 работает, Ctrl/ не работает, Ctrl. не работает (невозможно записать клавиатурный макрос). Но может это потому, что у меня оригинальный (не патченый) PuTTY.

да ванильный не годится :)

**Comment by sorcodiv** [comment]
Посмотрел на причины ошибки сборки в ubuntu 22.04. Проблема в версии libc.
В ubuntu 16.04 версия libc 2.23, в debian 11 версия libc 2.31, в ubuntu 22.04 версия libc 2.35.
Сборки, полученные в окружении со старой libc спокойно работают в новых ОС, а в обратную сторону - все сложнее.
Поэтому "из коробки" far2l-portable, собранный в ubuntu 22, не работает в debian 11.
Это, кстати, часто обсуждается, например https://stackoverflow.com/questions/2856438/how-can-i-link-to-a-specific-glibc-version
Статическая линковка libc не является решением: https://stackoverflow.com/questions/57476533/why-is-statically-linking-glibc-discouraged/
Есть проект, предназначенный для решения этой проблемы: https://github.com/wheybags/glibc_version_header
Но у меня не получилось: собранная версия far2l с заголовками force_link_glibc_2.23.h все равно ругается libc.so.6: version `GLIBC_2.32' not found (required by ./far2l)
Поэтому для сборки максимально совместимой версии приходится использовать Ubuntu 16.04.
В одном из обсуждений дана именно эта рекомендация: https://stackoverflow.com/questions/2856438/how-can-i-link-to-a-specific-glibc-version#62998901
Just build on the system with the oldest GLIBC you still want to support.

Добавил в скрипт установку нужных библиотек, сборка с USEWX=yes прошла успешно (в том числе SMB для NetRocks).
Проверял на manjaro-kde, manjaro-xfce, ubuntu-22.04-desktop - portable WX-версия запускается нормально.
[build.zip](https://github.com/shmuz/far2l/files/10550847/build.zip)

**Comment by shmuz** [comment]
О, это хорошие новости. Я, собственно, не вижу проблемы в использовании Ubuntu 16, просто было любопытно.
Кстати, почему бы вам не поместить эту работу в репозиторий? Это же очень удобно.

Уже пару дней вожусь с [PPA](https://launchpad.net/~far2l-team/+archive/ubuntu/lua), ребята выкладывают уже месяц или два, но видимо ни разу не проверили, что оно работает. У меня рабочий вариант получается только, если заменяю luafar2l.so на тот же файл из моей сборки.

P.S, Сегодня (1 февраля) совместными усилиями довели PPA до рабочего состояния.

**Comment by sorcodiv** [comment]
Разместил скрипты у себя: https://github.com/sorcodiv/far2lua-build
Как считаете, стоить выложить бинарные сборки?
И еще вопрос, в последней сборке после запуска far у меня в домашнем каталоге создается файл luafar_log.txt, текущую версию можно считать стабильной, или лучше подождать?

**Comment by shmuz** [comment]
Хорошо, что выложили, теперь у них есть свой дом :)
- Насчёт выкладывания бинарных сборок - можно, но без "рекламы" никто о них не узнает. Тем не менее, если готовая сборка будет лежать, то при случае, в форуме или в телеграме можно будет дать на неё ссылку.
- Насчёт luafar_log.txt - этот файл создаётся только коммитом от 30 января (мне нужно было дебажить PPA). После этого уже было 2 коммита, в них это убрано.

**Comment by shmuz** [comment]
По поводу стабильности: это понятие относительное, но это один из моих приоритетов.
Если бы было больше пользователей, безусловно всплыли бы проблемы, о которых я и не подозреваю. Тогда можно было бы ими заняться.

Не знаю, в курсе ли вы, но [здесь](https://github.com/shmuz/LuaFAR-2L) лежат несколько плагинов и несколько десятков готовых макросов для данного форка far2l.

**Comment by sorcodiv** [comment]
Бинарники нужны скорее тем, кто хочет просто посмотреть как работает приложение, без необходимости связываться со сборкой и без установки дополнительных пакетов. Думаю тот, кто действительно захочет опробовать far2l, найдет бинарники или соберет сам. А ссылка на это обсуждение есть на форуме farmanager.
Выложил сборку текущей версии: https://github.com/sorcodiv/far2lua-build/releases
Для проверки запустил сборку в актуальных debian, ubuntu, arch, manjaro kde/xfce, fedora, gentoo - с запуском проблем не нашел, базовые lua-макросы отрабатывают (не считая ситуаций, когда ОС перехватывает нажатия комбинаций клавиш типа ctrl+., но это совсем другая история).

**Comment by shmuz** [comment]
Сейчас попробовал ваш бинарник (который с WX) на Linux Mint.
Всё красиво, правда время запуска велико (20 секунд), но это видимо из-за загрузки всех библиотек.
Ещё меня удивило, что luafar2l.so, помимо своего обычного местонахождения (`$FARHOME/Plugins/luafar`), также находится и в папке `$FARHOME/lib`, причём размеры этих двух файлов разные.

Ещё попробовал на Debian 11.5 (32 bit). Тут он работать отказался, с ошибкой `Exec format error`. Видимо из-за разницы в битности.

**Comment by sorcodiv** [comment]
В процессе работы скрипт собирает все библиотеки, от которых зависит far2l и его плагины, в каталог ./lib/. С помощью [patchelf](https://github.com/NixOS/patchelf) во всех бинарных файлах производится подмена путей для поиска динамически подключаемых библиотек на внутренний каталог ./lib/. В том числе в плагине ./Plugins/luafar/luamacro/plug/luamacro.far-plug-wide. Чтобы не менять алгоритм работы скрипта, я копирую библиотеку luafar2l.so в./lib/, после этого во всех файлах производится подмена путей поиска динамических библиотек. Поэтому в ./lib/luafar2l.so поиск идет в $ORIGIN/../lib, а в ./Plugins/luafar/luafar2l.so поиск идет в $ORIGIN/../../lib - из-за этого размер отличается (отличий на самом деле больше, это один из примеров).

По поводу 32 bit - я выложил только сборку x64, поэтому в x32 она запускаться не будет. Добавил информацию в название файлов.

**Comment by shmuz** [comment]
Значит видимо скрипт должен в конце работы удалять (или не добавлять) `luafar/luafar2l.so`, т.к. она не используется.
Насчёт 32 бит - не думаю, что имеет смысл делать бинарники для таких систем. Я её поставил только, чтобы посмотреть как Фар работает на 32-битных системах.

**Comment by sorcodiv** [comment]
Спасибо, внес изменения в скрипт. И добавил x32-сборку.

**Comment by shmuz** [comment]
Попробовал сборку x32.
Стартует не за 20, а за 4 секунды (разница возможно из-за wayland).
Пришлось поиграться с выбором фонта, чтобы устраивал как мой существующий Фар, так и Фар из сборки.
Для многих фонтов один из фаров отображается ОК, другой - как будто есть пробел между смежными символами..

**Comment by sorcodiv** [comment]
Да, у меня в debian с gnome повторяется. Это связано с тем, что в сборку включены достаточно старые версии библиотек из ubuntu 16.
Убедиться в этом можно таким способом:
-установить в систему необходимые для far2l пакеты,
-распаковать сборку far-portable (это по сути tar.gz архив, но для распаковки надо произвести дополнительные действия, либо воспользоваться плагином arclite для windows-версии far),
-сравнить файлы (по названию) в распакованном каталоге ./lib/ с /usr/lib/i386-linux-gnu/,
-удалить из ./lib/ библиотеки, которые уже есть в системе.
Теперь far2l начнет загружать установленные в системе актуальные версии библиотек - после этого у меня шрифт, предложенный по умолчанию, начинает отображаться корректно.
С одной стороны, решением может стать исключение старых версий библиотек, включенных в portable-сборку.
Но с другой стороны, сборка станет не совсем portable - в каких-то случаях будет требовать предварительной установки пакетов в систему (а это уже требует прав root, которые есть не всегда).

**Comment by shmuz** [comment]
Можно сделать отдельные "профили" (сейчас это возможно только при запуске с ключом `-u`), и тогда фонт для portable будет независим от фонта для "стационарного" Фара. У меня на моём основном Linux Mint сделан ярлык с ключом `-u` для far2l от elfmz, и он имеет отдельную конфигурацию.

P.S. Нужно запускать так (в предположении, что назовём профиль `portable`):
`./far -- -u portable`

**Comment by shmuz** [comment]
@sorcodiv 
Проект переименован в far2m.
Соответственно изменился адрес: https://github.com/shmuz/far2m

**Comment by sorcodiv** [comment]
Для информации: немного изменил подход к собрке портативной версии, попробовал сделать ее чуть ближе к "стилю windows".
Добавил сборку некоторых плагинов из https://github.com/shmuz/luafar2m.
В имени файла указан codename дистрибутива, в котором осуществлялась сборка. Сборка бинарно-повторяема (во всяком случае у меня собранные в разное время файлы бинарно совпадают).
Для меня основным сценарием применения этой сборки явяляется использование на системах без прав root (или когда по другим причинам нет желания/возможности устанавливать недостающие пакеты в систему).
Сборка сделана для разных архитектур. Для уменьшения размера системные библиотеки вынесены в отдельный архив. Базово far будет работать и без системных билиотек. Доп. библиотеки выложены только для x64 сборки (как наиболее часто используемой). Для работы gui-плагина необходимо добавлять в каталог с far только недостающие в системе бибилиотеки (иначе может не работать gui, в зависимости от дистрибутива).
Выложил тут: https://github.com/sorcodiv/far2m-build/releases

**Comment by shmuz** [comment]
Наверное надо будет добавить в README.md ссылочку на эту репозиторию и ещё парочку других, наподобие.
Спасибо.

**Comment by shmuz** [comment]
Добавил в README.md.
